sql/core/src/main/scala/org/apache/spark/sql/execution/ExpandExec.scala

```diff
@@ -42,9 +42,27 @@ case class ExpandExec(
   override lazy val metrics = Map(
     "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"))
 
-  // The GroupExpressions can output data with arbitrary partitioning, so set it
-  // as UNKNOWN partitioning
-  override def outputPartitioning: Partitioning = UnknownPartitioning(0)
+  /**
+   * The Expand is commonly introduced by the RewriteDistinctAggregates optimizer rule.
+   * In that case there can be several attributes that are kept as they are by the Expand.
+   * If the child's output is partitioned by those attributes, then so will be
+   * the output of the Expand.
+   * In general case the Expand can output data with arbitrary partitioning, so set it
+   * as UNKNOWN partitioning.
+   */
+  override def outputPartitioning: Partitioning = {
+    val stableAttrs = ExpressionSet(output.zipWithIndex.filter {
+      case (attr, i) => projections.forall(_(i).semanticEquals(attr))
+    }.map(_._1))
+
+    child.outputPartitioning match {
```

Comment:
How about the `PartitioningCollection` case?

Reply:
Added that case, but I was not able to construct a test case for it

Reply:
we cannot use the shuffled join path for tests?  https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledJoin.scala#L48-L49

Reply:
The `ProjectExec`, that was inserted after join, managed to simplify it to `HashPartitioning` for the cases I could think of.

Reply:
Ah, I see.